oak/common/nonce_generator.cc

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "oak/common/nonce_generator.h"
+
+#include <openssl/err.h>
+#include <openssl/rand.h>
+
+#include "absl/strings/escaping.h"
+
+namespace oak {
+
+namespace {
+
+// Copied from https://github.com/google/asylo/blob/master/asylo/crypto/util/bssl_util.cc.
+std::string BsslLastErrorString() {
+  static constexpr int kErrorStringBufferLength = 256;
+  static char buffer[kErrorStringBufferLength];
+  ERR_error_string_n(ERR_get_error(), buffer, sizeof(buffer));
+  ERR_clear_error();
+  return std::string(buffer);
+}
+
+}  // namespace
+
+NonceGenerator::NonceGenerator(size_t nonce_size_bytes) : nonce_size_bytes_(nonce_size_bytes) {}
+
+// Copied from https://github.com/google/asylo/blob/master/asylo/crypto/random_nonce_generator.cc.
+asylo::StatusOr<Nonce> NonceGenerator::NextNonce() {
+  // Initialize nonce with |nonce_size_bytes_| zeros.
+  Nonce nonce(nonce_size_bytes_, 0);
+  if (RAND_bytes(nonce.data(), nonce.size()) != 1) {
+    return asylo::Status(asylo::error::GoogleError::INTERNAL,
+                         absl::StrCat("RAND_bytes failed: ", BsslLastErrorString()));
```

Comment:
For `oak.random_get` I used C++'s `std::random_device`, which has the advantage of not needing to cope with an error return (and not needing an OpenSSL dependency).  Not sure that was the right choice, but maybe we should use a consistent mechanism?

Reply:
Good suggestion! Using that now, PTAL.